### PR TITLE
Upgrade to Hive 3.0.0

### PR DESCRIPTION
**Pull request by imjalpreet**
This PR includes the base changes required for upgrading to Hive 3. This will help in supporting and bringing a number of interesting features in Hive 3 to Presto in the future.

This depends on https://github.com/prestodb/presto-hive-apache/releases/tag/3.0.0-2 and the following PR in tempto (https://github.com/prestodb/tempto/pull/272) 

depends on https://github.com/facebookexternal/presto-facebook/pull/1414

### CODE REVIEW

**Comment by aweisberg** [comment]
If there are test failures introduced by a commit it's better to have the fixes in the commit and not separate. Otherwise there are failing commits on master that interfere with bisecting.

This seems to have introduced another dependency related failure. It's caught in the FB integration builds.
```
[WARNING] Rule 0: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for com.facebook.presto.hive:hive-apache:1.2.2-2 paths to dependency are:
+-com.facebook.presto:presto-namespace:0.248-SNAPSHOT
  +-com.facebook.presto:presto-hive-metastore:0.248-SNAPSHOT
    +-com.facebook.presto.hive:hive-apache:1.2.2-2 (managed) <-- com.facebook.presto.hive:hive-apache:3.0.0-2
, 
Require upper bound dependencies error for org.apache.thrift:libthrift:0.9.1 paths to dependency are:
+-com.facebook.presto:presto-namespace:0.248-SNAPSHOT
  +-com.facebook.presto:presto-hive-metastore:0.248-SNAPSHOT
    +-org.apache.thrift:libthrift:0.9.1 (managed) <-- org.apache.thrift:libthrift:0.9.3
]
```

**Comment by imjalpreet** [comment]
@aweisberg Yes once the tempto changes are through I will update this PR and make sure all the test cases pass before the code is reviewed. 

Regarding the Facebook Integration build failure, it looks like there are some additional plugins in FB like presto-namespace which will also have to be modified for Hive 3 Upgrade. Currently, this additional plugin is using some dependencies which have been upgraded in this PR due to which we are seeing the errors. I am not sure if this plugin (presto-namespace) is open-sourced.

**Comment by aweisberg** [comment]
Thanks I missed that. I thought presto-namespace was one of the OSS repos. Still trying to figure out how to release Tempto.

**Comment by aweisberg** [comment]
So the issue is that we can't release Tempto because it's under `io.prestodb` which isn't controlled by the Linux Foundation.

We either need to find out who has the credentials for updating `io.prestodb` and get them to provide access or we need to refactor Tempto so we can upload it to `com.facebook.presto`.

I can ask around about updating `io.prestodb` but it could take a while, and I can't guarantee it will happen. I broached the subject of refactoring it under `com.facebook.presto` and so far no one has objected.

If you want to do the refactor of Tempto it would unblock this PR otherwise we have to wait on trying to get the credentials back. Sorry we are getting blocked on something so basic.

**Comment by imjalpreet** [comment]
@aweisberg Thanks for looking into this. I can take up the refactoring task and raise another PR once it's ready.

**Comment by aweisberg** [comment]
Great, so the package we want is `com.facebook.prestodb.tempto` and `find ./ -type f -exec sed -i '' -e 's/io.prestodb/com.facebook.prestodb/g' {} \;` can replace anything that refactoring in IntelliJ doesn't manage to do.

**Comment by aweisberg** [comment]
[Asking Sonatype if we can get access to `io.prestodb.tempto`](https://issues.sonatype.org/browse/OSSRH-37091?focusedCommentId=1015692&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-1015692)

**Comment by aweisberg** [comment]
I was able to upload the snapshot artifacts. Can you update and rebase your PR to use the 1.51-SNAPSHOT artifact and we will see if the tests complete?

**Comment by imjalpreet** [comment]
@aweisberg I had made the respective changes. The product tests have passed but the tests for presto-main module seem to have failed. It doesn't look like that is because of the code changes in this PR, looks like it was because of some crash while the tests were running. Maybe if you can re-trigger the module, it would pass successfully.

**Review by aweisberg — CHANGES_REQUESTED** [review]
Only two minor change requests. The next step is for me to deploy this on a cluster and shadow some traffic to sanity check. I don't want this to get reverted during the release process.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/BloomFilter.java
Comment (imjalpreet) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/BloomFilter.java:1]: @aweisberg The latest version was not taken with the idea of making the changes gradually to understand the impact. The class being used in trino is forked from org.apache.orc.util.BloomFilter and we had been referring to org.apache.hive.common.util.BloomFilter.

Let me check the impact and the changes required and get back.

Comment (imjalpreet) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/BloomFilter.java:1]: @aweisberg I reviewed the BloomFilter Class being used in Trino. As I mentioned earlier, one of the major change is they have forked from **org.apache.orc.util.BloomFilter** and post that they have made some optimisations on top of that. A couple of enhancements and optimisations include changes in ORC Bloom filter hash, added support for ORC UTF8 bloom filters, real type in ORC bloom filter and a few more. 

Would it be better to have separate PRs for these enhancements?

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java
Comment (imjalpreet) [reply presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java]: @aweisberg I have made the refactor consistent with Trino as you requested.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/BloomFilter.java
Comment (aweisberg) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/statistics/BloomFilter.java:1]: I would have been fine with that as well :-) Thanks for doing it now. I am still trying to figure out why I can't publish Tempto. It really does seem to be uploading without error.

**Comment by aweisberg** [comment]
It's released! At long last our national nightmare is over.

Bad news, there are some shading issues I think. FB integration is failing on `duplicate-finder-maven-plugin` which I think is finding legitimate issues even outside the FB integration. We should probably enable the duplicate finder on Presto Hive https://pastebin.com/AHxAnJNF

**Comment by imjalpreet** [comment]
> It's released! At long last our national nightmare is over.
> 
That's great news.
> 
> 
> Bad news, there are some shading issues I think. FB integration is failing on `duplicate-finder-maven-plugin` which I think is finding legitimate issues even outside the FB integration. We should probably enable the duplicate finder on Presto Hive https://pastebin.com/AHxAnJNF 
> 
Sure I will also try to have a look into the same.

**Review by aweisberg — APPROVED** [review]
LGTM. I verified this on a cluster. Turns out the issue was that the Tempto 1.51-SNAPSHOT was probably just unmodified 1.50 since I was failing it update the artifacts. Switching to the 1.51 release artifact made the dependency issues go away.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java
Comment (rschlussel) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java:826]: What's the reason for the changes for decimal?

File: presto-raptor/src/test/java/com/facebook/presto/raptor/storage/TestShardWriter.java
Comment (rschlussel) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/storage/TestShardWriter.java:255]: why was this test removed?

**Review by rschlussel — COMMENTED** [review]
Just 2 questions, otherwise looks good.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java
Comment (aweisberg) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java:826]: Constructor argument order is reversed in Hive 3. Trino also just switches the parameters around.

File: presto-raptor/src/test/java/com/facebook/presto/raptor/storage/TestShardWriter.java
Comment (aweisberg) [reply presto-raptor/src/test/java/com/facebook/presto/raptor/storage/TestShardWriter.java:255]: Short answer, because that is what Trino did https://github.com/trinodb/trino/commit/238f4b6bb1ac98130cff7bee88d5e622e5a2808f#diff-82a51d3cf239ee341a4eab63c8c88c82367b389a9652cbaa067631d131a60930

Long answer... I don't know. I'll bring it back and see what happens when I run it.

Comment (aweisberg) [reply presto-raptor/src/test/java/com/facebook/presto/raptor/storage/TestShardWriter.java:255]: The test still passes. So I guess we should keep it!

**Review by aweisberg — CHANGES_REQUESTED** [review]
@imjalpreet can you bring back `TestShardWriter.testWriterZeroRows`? I am not sure why it was removed in Trino, but I brought it back and the test still passes.

**Comment by imjalpreet** [comment]
@aweisberg I have reverted that test removal as requested.

**Review by rschlussel — APPROVED** [review]

**Review by aweisberg — APPROVED** [review]
Thank you @imjalpreet

**Comment by rschlussel** [comment]
Ignoring travis failure since we're migrating from travis.  Thanks for the contribution!

**Comment by aweisberg** [comment]
Unfortunately this seems to have broken a query in timezone 'America/Dawson'. We have an instance where it is returning different results after the upgrade. Haven't root caused why exactly yet, but I hope to be able to share a reproducer soon.

**Comment by aweisberg** [comment]
```
          (
            (
              (active.ds = '2021-03-01') 
              AND (
                time_stop > "to_unixtime"(
                  "date_trunc"(
                    'DAY', 
                    "parse_datetime"(
                      "concat"(
                        '2021-03-01 06:00:00 ', lookup.timezone_name
                      ), 
                      'yyyy-MM-dd HH:mm:ss ZZZ'
                    )
                  )
                )
              )
            ) 
            AND (
              time_start < "to_unixtime"(
                "date_trunc"(
                  'DAY', 
                  "parse_datetime"(
                    "concat"(
                      '2021-03-02 06:00:00 ', lookup.timezone_name
                    ), 
                    'yyyy-MM-dd HH:mm:ss ZZZ'
                  )
                )
              )
            )
          )
```
The predicate that is inconsistent, Lookup.timezone_name is always 'America/Dawson'. time_stop is either 0 or a value like 1615100879 and one example is of time start (paired with the non-zero time stop) is 1614668884.

https://pastebin.com/FjemTqE4 is the values. Every instance of these was present in 0.227, but was not materialized by the query in 0.228.

It's also possible it's not actually this predicate, but the grouping being different on these two columns:
```
          IF(
            (
              active.time_stop > "to_unixtime"(
                "date_trunc"(
                  'DAY', 
                  "parse_datetime"(
                    "concat"(
                      '2021-03-02 06:00:00 ', lookup.timezone_name
                    ), 
                    'yyyy-MM-dd HH:mm:ss ZZZ'
                  )
                )
              )
            ), 
            "to_unixtime"(
              "date_trunc"(
                'DAY', 
                "parse_datetime"(
                  "concat"(
                    '2021-03-02 06:00:00 ', lookup.timezone_name
                  ), 
                  'yyyy-MM-dd HH:mm:ss ZZZ'
                )
              )
            ), 
            active.time_stop
          ) time_stop_daily, 
          IF(
            (
              active.time_start < "to_unixtime"(
                "date_trunc"(
                  'DAY', 
                  "parse_datetime"(
                    "concat"(
                      '2021-03-01 06:00:00 ', lookup.timezone_name
                    ), 
                    'yyyy-MM-dd HH:mm:ss ZZZ'
                  )
                )
              )
            ), 
            "to_unixtime"(
              "date_trunc"(
                'DAY', 
                "parse_datetime"(
                  "concat"(
                    '2021-03-01 06:00:00 ', lookup.timezone_name
                  ), 
                  'yyyy-MM-dd HH:mm:ss ZZZ'
                )
              )
            ), 
            active.time_start
          ) time_start_daily 
```

**Comment by aweisberg** [comment]
I should add it cuts both ways and there are values (215) that appeared in 0.228 that didn't appear in 0.227

**Comment by aweisberg** [comment]
0.247 
```
presto:ad_delivery> select "to_unixtime"( "date_trunc"( 'DAY', "parse_datetime"( "concat"( '2021-03-02 06:00:00 ', 'America/Dawson' ), 'yyyy-MM-dd HH:mm:ss ZZZ' ) ));
   _col0              
------------
 1.614672E9 
(1 row)
```

0.248
```
presto:ad_delivery> select "to_unixtime"( "date_trunc"( 'DAY', "parse_datetime"( "concat"( '2021-03-02 06:00:00 ', 'America/Dawson' ), 'yyyy-MM-dd HH:mm:ss ZZZ' ) ));
    _col0             
-------------
 1.6146684E9 
(1 row)
```

So they wouldn't group the same way, but this also suggests that the predicates will be slightly off as well. These are off by an hour.

**Comment by aweisberg** [comment]
I was mistaken the joda upgrade was in https://github.com/prestodb/presto/pull/15649, hah I should know!

**Comment by imjalpreet** [comment]
> I was mistaken the joda upgrade was in https://github.com/prestodb/presto/pull/15649, hah I should know!

Ah okay, no issues!

**Comment by aweisberg** [comment]
This did end up failing on at least one different issue. CONTAINS on an array was returning false and never true when it should and the # of input rows counted by the aggregation for that query also didn't match 0.247.

I am trying to factor out a reproducer now.

**Comment by aweisberg** [comment]
Still can't reproduce, but I noticed that the format of the table that creates this issue is TEXTFILE. I can create my own TEXTFILE format table with an array it in, and select it out, but I don't see that the values in the array being merged :-(

**Comment by aweisberg** [comment]
LOL https://forums.aws.amazon.com/thread.jspa?threadID=264065

**Comment by aweisberg** [comment]
https://github.com/trinodb/trino/pull/1733/files
https://github.com/trinodb/trino/pull/1321/files

**Comment by imjalpreet** [comment]
@aweisberg thanks for the detailed investigation. A bug due to correcting a misspelled property that's a first. I will try to get in the fix before the release cut most probably by today evening.

**Comment by highker** [comment]
Thanks @imjalpreet putting this up. While working on the fix, please take a look at #15589. It depends which PR merges first. We will need to add materialized view support for Hive 3.0

**Comment by imjalpreet** [comment]
@highker Sure, I will have a look at the materialized view PR.

**Comment by aweisberg** [comment]
@highker I am not 100% following. Do we need to deliver both at the same time? If we only release the Hive 3 upgrade is that ok?

**Comment by highker** [comment]
@aweisberg, @imjalpreet, for example, we need to model materialized views as managed tables today (https://github.com/prestodb/presto/pull/15589#discussion_r592098962). This can be changed after Hive 3.0

**Pull request merged**
